Algebraic-Diagrammatic-Construction/EE_ADC2.py

```diff
@@ -0,0 +1,147 @@
+'''
+A reference implementation of ADC(2) for the calculation of excited
+states for a restricted Hartree-Fock reference. A spin orbital
+formulation is used, as it simplifies the equations.
+
+References:
+    M. Hodecker, PhD Thesis, Heidelberg University (2020).
+'''
+
+__authors__ = 'Oliver J. Backhouse'
+__credits__ = ['Oliver J. Backhouse']
+
+__copyright__ = '(c) 2014-2020, The Psi4NumPy Developers'
+__license__ = 'BSD-3-Clause'
+__date__ = '2018-03-01'
+
+import time
+import numpy as np
+import psi4
+import functools
+from adc_helper import davidson
+
+einsum = functools.partial(np.einsum, optimize=True)
+
+
+# Settings
+n_states = 5
+tol = 1e-8
+
+# Set the memory and output file
+psi4.set_memory('2 GB')
+psi4.core.set_output_file('output.dat', False)
+
+# Set molecule and basis
+mol = psi4.geometry('''
+O
+H 1 1.1
+H 1 1.1 2 104
+symmetry c1
+''')
+
+psi4.set_options({'basis': '6-31g',
+                  'scf_type': 'pk',
+                  'mp2_type': 'conv',
+                  'e_convergence': 1e-10,
+                  'd_convergence': 1e-10,
+                  'freeze_core': 'false'})
+
+# Perform SCF
+print('\nPerforming SCF...')
+e_scf, wfn = psi4.energy('SCF', return_wfn=True)
+mints = psi4.core.MintsHelper(wfn.basisset())
+
+# Get data from the wavefunction
+nocc = wfn.doccpi()[0]
+nmo = wfn.nmo()
+e_mo = wfn.epsilon_a().np
+c_mo = wfn.Ca()
+
+# Get the antisymmetrized spin-orbital integrals (physicist notation)
+print('Building integrals...')
+eri = mints.mo_spin_eri(c_mo, c_mo).np
+
+# Expand to spin orbitals
+nso = nmo * 2
+nocc = nocc * 2
+nvir = nso - nocc
+nia = nocc * nvir
+e_mo = np.repeat(e_mo, 2)
+
+# Build some slices
+o = slice(None, nocc)
+v = slice(nocc, None)
+
+# Calculate intermediates
+e_ia = e_mo[o,None] - e_mo[None,v]
+e_ijab = e_ia[:,None,:,None] + e_ia[None,:,None,:]
+t2 = eri[o,o,v,v] / e_ijab
+
+# Print the MP2 energy
+e_mp2 = einsum('ijab,ijab->', t2, eri[o,o,v,v]) * 0.25
+print('RHF total energy:       %16.10f' % e_scf)
+print('MP2 correlation energy: %16.10f' % e_mp2)
+print('MP2 total energy:       %16.10f' % (e_scf+e_mp2))
+psi4.compare_values(psi4.energy('mp2'), e_mp2+e_scf, 6, 'MP2 Energy')
+
+# Define the operation representing the dot-product of the EE-ADC(2) matrix
+# with an arbitrary state vector.
+def matvec(y):
```

Comment:
Can you link to equations int he paper here? Same request for your other matvec functions.

Reply:
Done in https://github.com/psi4/psi4numpy/pull/105/commits/40d8c84b67d175f098f17bce9926af6af3b372d4